### PR TITLE
Improve maximizeLogLikelihood algorithm in MastodonUtils.C

### PR DESCRIPTION
**Pull request by somu15**
Developed a Stochastic Gradient Descent (SGD) function in MastodonUtils.C to find log-normal parameters that fit fragility data. Users now have two options to fit fragility functions: BRUTE FORCE or SGD. Also, ensured that SGD implementation stays within the parameter bounds specified by the user.

### CODE REVIEW

**Comment by cbolisetti** [comment]
@somu15 your test is failing. Please check.

File: doc/content/source/vectorpostprocessors/Fragility.md [hunk 1]
```diff
@@ -12,6 +12,14 @@ demand distribution and capacity distribution and calculate the conditional prob
 in each bin, and 3) fit a lognormal distribution in the conditional probabilities to calculate a
 fragility. This kind of fragility is also referred to as 'enhanced fragility'.
 
+Fitting a Log-Normal fragility function given analysis results requires optimization to find
```
[the remark this answers is not in the document]
Comment (somu15) [reply doc/content/source/vectorpostprocessors/Fragility.md]: Changed this.

File: doc/content/source/vectorpostprocessors/Fragility.md [hunk 2]
```diff
@@ -12,6 +12,14 @@ demand distribution and capacity distribution and calculate the conditional prob
...
+the parameters location and scale. Two optimization methods are currently available: Brute Force
+and Stochastic Gradient Descent (SGD). Brute Force finds that combination of parameters
+ which maximizes the likelihood by looping over all alternative combinations. SGD finds the parameter
+ combination by running a Newton-Raphson type algorithm with random seeds. Of course, both algorithms
+ have their pros and cons. For example, Brute Force takes orders of magnitude more time to run. SGD
+ does not always guarantee precise results.
```
[the remark this answers is not in the document]
Comment (somu15) [reply doc/content/source/vectorpostprocessors/Fragility.md]: Implemented these changes, thank you.

File: doc/content/source/vectorpostprocessors/Fragility.md [hunk 3]
```diff
@@ -21,13 +29,15 @@ that are sensitive to floor spectral acceleration demands at a particular freque
...
 
-An example usage of the `Fragility` VectorPostprocessor is below. Apart from the input file,
-`fragility.i` the this VectorPostprocessor requires the csv files containing the output data from
-the probabilistic or stochastic simulations. An example output data file, `master_out_run_hazard03_sub1.csv`,
-is also shown below. Note that the inputs to the parameters `master_file`, `hazard_multiapp`, and
-`probabilistic_multiapp` are used to build the filename of the csv files that contain the results of
-probabilistic simulations. Additionally, the `demand_variable` input parameter (accel_2x) in this case
-is identical to the column name in the output data file.
+An example usage of the `Fragility` VectorPostprocessor is below. For the `optimization_method`
+optimization_method option, either BRUTE FORCE or SGD (i.e., Stochastic Gradient Descent) can
```
[the remark this answers is not in the document]
Comment (somu15) [reply doc/content/source/vectorpostprocessors/Fragility.md]: Removed it, thank you.

File: include/utils/MastodonUtils.h [hunk 4]
```diff
@@ -145,11 +145,30 @@ Real calcLogLikelihood(const std::vector<Real> & im,
...
+                                        const std::vector<Real> & loc_space,
+                                        const std::vector<Real> & sca_space,
+                                        const unsigned int & n,
+                                        const std::string & method,
+                                        const Real tolerance,
+                                        const Real gamma,
+                                        const int num_rnd,
+                                        const int seed);
+
+/*std::vector<Real> maximizeLogLikelihood_Crude(const std::vector<Real> & im,
```
[the remark this answers is not in the document]
Comment (somu15) [reply include/utils/MastodonUtils.h]: Removed those, thank you.

**Comment by cbolisetti** [comment]
@somu15 looks like you didn't push your changes. Please push them and I'll merge this soon.

**Comment by somu15** [comment]
@cbolisetti I will push it as soon as the unit tests pass. I haven't unit tested the code with new changes.

File: doc/content/source/vectorpostprocessors/Fragility.md [hunk 5]
```diff
@@ -12,6 +12,22 @@ demand distribution and capacity distribution and calculate the conditional prob
 in each bin, and 3) fit a lognormal distribution in the conditional probabilities to calculate a
 fragility. This kind of fragility is also referred to as 'enhanced fragility'.
 
+Fitting a lognormal fragility function given analysis results requires optimization to find
```
Comment (cbolisetti) [inline doc/content/source/vectorpostprocessors/Fragility.md]: Please change this to:
Fitting a lognormal fragility function to the probabilities of failure requires an optimization algorithm to find the parameters, median and lognormal standard deviation. Two methods are currently available for calculating these fragility parameters: 'brute force' and 'Stochastic Gradient Descent (SGD)'.

File: doc/content/source/vectorpostprocessors/Fragility.md [hunk 6]
```diff
@@ -12,6 +12,22 @@ demand distribution and capacity distribution and calculate the conditional prob
...
+ which maximizes the likelihood by looping over all alternative combinations. SGD finds the parameter
+ combination by running a Newton-Raphson type algorithm with random seed values. SGD also requires the additional parameters:
+
+1. tolerance (To declare convergence of the algorithm. 1e-03 can be optimal.)
+2. gamma (Controls the step size of the algorithm. 0.001 can be optimal.)
+3. numrnd (Number of random seed values. 1000 can be optimal.)
+4. seed (Seed for the random number generator. Currently set to 1028).
+
+Of course, both algorithms have their pros and cons. For example, brute force takes orders of magnitude more time to run. SGD
+does not always guarantee precise results.
```
Comment (cbolisetti) [inline doc/content/source/vectorpostprocessors/Fragility.md]: Please say that 'SGD may not guarantee the best fragility fit and users should try to change the SGD parameters to ensure that the SGD results do not change significantly.'

File: doc/content/source/vectorpostprocessors/Fragility.md [hunk 7]
```diff
@@ -12,6 +12,22 @@ demand distribution and capacity distribution and calculate the conditional prob
...
+
+Additionally, there is a input variable called 'num_collapses' which is used to calculate likelihood when using Baker's MLE.
```
Comment (cbolisetti) [inline doc/content/source/vectorpostprocessors/Fragility.md]: 'an input variable'.

File: unit/src/TestMastodonUtils.C [hunk 8]
```diff
@@ -316,11 +316,17 @@ TEST(MastodonUtils, maximizeLogLikelihood)
   std::vector<Real> loc_space = {0.2, 0.6};
   std::vector<Real> sca_space = {0.1, 0.4};
   unsigned int n = 1000;
+  bool method1 = true;
+  bool method2 = false;
   // Outputs for testing
-  std::vector<Real> max_values =
-      MastodonUtils::maximizeLogLikelihood(im, pf, loc_space, sca_space, n);
+  std::vector<Real> max_values1 = MastodonUtils::maximizeLogLikelihood(
+      im, pf, loc_space, sca_space, n, method1, 1e-05, 0.0001, 100, 1028);
```
Comment (cbolisetti) [inline unit/src/TestMastodonUtils.C]: please remove method1 and method2 and use true/false in the function call

File: unit/src/TestMastodonUtils.C [hunk 9]
```diff
@@ -316,11 +316,17 @@ TEST(MastodonUtils, maximizeLogLikelihood)
   std::vector<Real> loc_space = {0.2, 0.6};
   std::vector<Real> sca_space = {0.1, 0.4};
   unsigned int n = 1000;
```
Comment (cbolisetti) [inline unit/src/TestMastodonUtils.C]: Please remove n here as well, and just use 1000 in the function call.

File: unit/src/TestMastodonUtils.C [hunk 10]
```diff
@@ -316,11 +316,17 @@ TEST(MastodonUtils, maximizeLogLikelihood)
...
+  std::vector<Real> max_values1 = MastodonUtils::maximizeLogLikelihood(
+      im, pf, loc_space, sca_space, n, method1, 1e-05, 0.0001, 100, 1028);
+  std::vector<Real> max_values2 = MastodonUtils::maximizeLogLikelihood(
+      im, pf, loc_space, sca_space, n, method2, 1e-05, 0.0001, 100, 1028);
   // Value check
-  EXPECT_TRUE(MooseUtils::absoluteFuzzyEqual(max_values[0], 0.4));
-  EXPECT_TRUE(MooseUtils::absoluteFuzzyEqual(max_values[1], 0.24));
+  EXPECT_TRUE(MooseUtils::absoluteFuzzyEqual(max_values1[0], 0.4));
+  EXPECT_TRUE(MooseUtils::absoluteFuzzyEqual(max_values1[1], 0.24));
+  EXPECT_TRUE(MooseUtils::absoluteFuzzyEqual((int)(max_values2[0] * 10000.0) / 10000.0, 0.3981));
```
Comment (cbolisetti) [inline unit/src/TestMastodonUtils.C]: Please change this to use the tolerance parameter in the absoluteFuzzyEqual function. Refer to the function definition in MooseUtils.h.

File: test/tests/vectorpostprocessors/fragility/tests [hunk 11]
```diff
@@ -5,10 +5,18 @@
...
   [./fragility]
     type = CSVDiff
-    input = fragility.i
-    csvdiff = fragility_out_fragility_pump_0002.csv
+    input = fragility_brute_force.i
+    csvdiff = fragility_brute_force_out_fragility_pump_0002.csv
     boost = true
 
-    requirement = "The Fragility vectorpostprocessor shall accurately evaluate the median demands, beta, and the conditional probability of failure of the SSC at each intensity, and the median and beta of the enhanced fragility fit for the SSC."
+    requirement = "The Fragility vectorpostprocessor shall accurately evaluate the median demands, beta, and the conditional probability of failure of the SSC at each intensity, and the median and beta of the enhanced fragility fit for the SSC using the stochastic gradient descent algorithm using a brute force approach."
```
Comment (cbolisetti) [inline test/tests/vectorpostprocessors/fragility/tests]: This requirement says '.. using a stochastic gradient descent algorithm using a brute force approach...'

File: doc/content/source/vectorpostprocessors/Fragility.md [hunk 12]
```diff
@@ -21,15 +37,16 @@ that are sensitive to floor spectral acceleration demands at a particular freque
...
 
-An example usage of the `Fragility` VectorPostprocessor is below. Apart from the input file,
-`fragility.i` the this VectorPostprocessor requires the csv files containing the output data from
-the probabilistic or stochastic simulations. An example output data file, `master_out_run_hazard03_sub1.csv`,
-is also shown below. Note that the inputs to the parameters `master_file`, `hazard_multiapp`, and
-`probabilistic_multiapp` are used to build the filename of the csv files that contain the results of
-probabilistic simulations. Additionally, the `demand_variable` input parameter (accel_2x) in this case
-is identical to the column name in the output data file.
+An example usage of the `Fragility` VectorPostprocessor is below. For the `brute_force`
+option, either true or false can be chosen. Apart from the input file, `fragility_brute_force.i`, the this VectorPostprocessor requires the
```
Comment (cbolisetti) [inline doc/content/source/vectorpostprocessors/Fragility.md]: rewrite this sentence

File: src/utils/MastodonUtils.C [hunk 13]
```diff
@@ -343,27 +344,90 @@ MastodonUtils::calcLogLikelihood(const std::vector<Real> & im,
...
+                                     const Real tolerance,
+                                     const Real gamma,
+                                     const int num_rnd,
+                                     const int seed)
 {
-  Real loglikelihoodmax = MastodonUtils::calcLogLikelihood(im, pf, loc_space[0], sca_space[0], n);
-  std::vector<Real> max_values = {0.0, 0.0};
-  for (Real loc = loc_space[0]; loc < loc_space[1]; loc += 0.01)
+  std::vector<Real> params_return = {0, 0};
+  if (brute_force == true)
```
Comment (cbolisetti) [inline src/utils/MastodonUtils.C]: Change this to
if (brute_force)

File: doc/content/source/vectorpostprocessors/Fragility.md [hunk 14]
```diff
@@ -12,6 +12,27 @@ demand distribution and capacity distribution and calculate the conditional prob
...
+2. gamma (Controls the step size of the algorithm. 0.001 can be optimal.)
+3. numrnd (Number of random seed values. 1000 can be optimal.)
+4. seed (Seed for the random number generator. Currently set to 1028).
+
+Of course, both algorithms have their pros and cons. For example, brute force takes orders
+of magnitude more time to run. SGD may not guarantee the best fragility fit and users should
+change the SGD parameters and make sure that the results don't change significantly, to
+ensure that the best results are calculated.
+
+Additionally, there is an input variable called 'num_collapses' which is used to calculate likelihood
```
Comment (cbolisetti) [inline doc/content/source/vectorpostprocessors/Fragility.md]: This sentence is out of place here. Please remove it.

**Pull request merged**